nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,344 @@
+  double origin_y = filter_mask_->info.origin.position.y;
+  double resolution = filter_mask_->info.resolution;
+  unsigned int size_x = filter_mask_->info.width;
+  unsigned int size_y = filter_mask_->info.height;
+
+  if (wx < origin_x || wy < origin_y) {
+    return false;
+  }
+
+  mx = std::round((wx - origin_x) / resolution);
+  my = std::round((wy - origin_y) / resolution);
+  if (mx >= size_x || my >= size_y) {
+    return false;
+  }
+
+  return true;
+}
+
+inline int8_t SpeedFilter::getMaskData(
+  const unsigned int mx, const unsigned int my) const
+{
+  return filter_mask_->data[my * filter_mask_->info.width + mx];
+}
+
+void SpeedFilter::process(
+  nav2_costmap_2d::Costmap2D & /*master_grid*/,
+  int /*min_i*/, int /*min_j*/, int /*max_i*/, int /*max_j*/,
+  const geometry_msgs::msg::Pose2D & pose)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      logger_, *(clock_), 2000,
+      "SpeedFilter: Filter mask was not received");
+    return;
+  }
+
+  geometry_msgs::msg::Pose2D mask_pose;  // robot coordinates in mask frame
+
+  // Transforming robot pose from current layer frame to mask frame
+  if (!transformPose(pose, mask_pose)) {
+    return;
+  }
+
+  // Converting mask_pose robot position to filter_mask_ indexes (mask_robot_i, mask_robot_j)
+  unsigned int mask_robot_i, mask_robot_j;
+  if (!worldToMask(mask_pose.x, mask_pose.y, mask_robot_i, mask_robot_j)) {
+    return;
+  }
+
+  // Getting filter_mask data from cell where the robot placed and
+  // calculating speed limit value
+  int8_t data = getMaskData(mask_robot_i, mask_robot_j);
+  if (data != SPEED_MASK_NO_LIMIT && data != SPEED_MASK_UNKNOWN) {
+    speed_limit_ = data * multiplier_ + base_;
+    if (percentage_) {
+      if (speed_limit_ < 0.0) {
+        RCLCPP_WARN(
+          logger_,
+          "SpeedFilter: Speed limit in filter_mask[%i, %i] is less than 0%, "
+          "which can not be true. Setting it to 0% value.",
+          mask_robot_i, mask_robot_j);
+        speed_limit_ = NO_SPEED_LIMIT;
+      }
+      if (speed_limit_ > 100.0) {
+        RCLCPP_WARN(
+          logger_,
+          "SpeedFilter: Speed limit in filter_mask[%i, %i] is higher than 100%, "
+          "which can not be true. Setting it to 100% value.",
+          mask_robot_i, mask_robot_j);
+        speed_limit_ = 100.0;
+      }
+    }
+  } else {
+    // Corresponding filter mask cell is free or unknown.
+    // Setting no speed limit there.
+    speed_limit_ = NO_SPEED_LIMIT;
+  }
+
+  if (speed_limit_ != speed_limit_prev_) {
+    if (speed_limit_ != NO_SPEED_LIMIT) {
+      if (percentage_) {
+        RCLCPP_DEBUG(
+          logger_,
+          "SpeedFilter: Speed limit is set to %f of maximum speed",
+          speed_limit_);
+      }
+    } else {
+      RCLCPP_DEBUG(logger_, "SpeedFilter: Speed limit is set to its default value");
```

Comment:
I think this log can be removed, if we change, then the other debug will trigger with a `100%` in this case. Just remove the second nested if statement, the difference in the semantics of the debug log isn't really worth the additional logic.

Reply:
> if we change, then the other debug will trigger with a 100% in this case

Unfortunately, due to our design this will be triggered to `0%` in this case and we will get "SpeedFilter: Speed limit is set to 0% of maximum speed" messages when speed is returning to its normal value. So, there is additional message for that case was added.

Reply:
OK - this block of code just seems a little much just to produce a log.

Reply:
Removed (temporarily) unnecessary `percent_` and extra check. This should look a little better in https://github.com/ros-planning/navigation2/commit/34e694dbee7af9f805d2a7038b292466e44dc104#diff-d344e376ecbc8e5ecd5045731e13855c6d2a9a955cdaee9b31dddbbee7681466L296 change